user-guide/Troubleshooting/Procedures/SLScripting_FinalizerException.md

```diff
@@ -0,0 +1,259 @@
+
+> [!NOTE]
+>
+> - Typically, you can avoid having to provide a finalizer by using the [SafeHandle](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle) or derived classes to wrap any unmanaged handle.
+> - We recommend providing a way to explicitly release resources before the garbage collector frees the object by implementing the [IDisposable](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.8.1) interface to perform the necessary cleanup for the object. This can considerably improve the performance of the application. Note, however, that even if you implement IDisposable, the finalizer acts as a safeguard to clean up the resources if the call to the Dispose method fails.
+
+> [!TIP]
+> For more information about cleaning up resources, refer to:
+>
+> - [Cleaning Up Unmanaged Resources](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/unmanaged)
+> - [Implementing a Dispose Method](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose)
+> - [Implementing a DisposeAsync Method](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync)
+> - [using statement](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/using)
+
+## Investigating the crash dump in Visual Studio
+
+In case a crash occurs because of an exception in the Finalize method of a class instance, a dump of SLManagedScripting should be created. The *ERRORLOG.TXT* file should contain an indication of this exception. For example:
+
+```txt
+System.NullReferenceException: Object reference not set to an instance of an object.
+   at Skyline.Protocol.MyExtension.MyClass.Finalize()
+```
+
+1. Open the SLManagedScripting dump file in Visual Studio, and click the *Run Diagnostic Analysis* link in the Actions list.
+
+   The *Analysis results* pane should show that an exception was found on the heap:
+
+   ![Analysis results](~/user-guide/images/VisualStudioDiagnosticAnalysFinalizerCrash.png)<br>*Visual Studio Analysis results*
+
+   In this case, it mentions that a [NullReferenceException](https://learn.microsoft.com/en-us/dotnet/api/system.nullreferenceexception?view=netframework-4.8.1) was found.
+
+1. Open the *Parallel Stacks* window (*Debug* > *Windows* > *Parallel Stacks*):
+
+   ![Visual Studio Parallel Stacks window](~/user-guide/images/VisualStudioParallelStacksWindowFinalizerCrash.png)<br>*Visual Studio Parallel Stacks window*
+
+   This shows that an unhandled exception occurred on the Finalizer thread (`OnUnhandledException`) while executing the Finalizer of MyClass (`~MyClass`).
+
+1. Inspect the source of the class to find out why the exception was thrown.
+
+## Investigating the crash dump with WinDbg
+
+Open the dump file with WinDbg.
+
+In this case, executing the `.lastevent` and `!analyze` commands do not reveal any useful info.
```

Comment:
Do you mean that these commands will never reveal useful info in the context of this investigation? For a user who isn't used to WinDbg, this will probably not be clear.

Reply:
depends on what it can find in the dump. I just mentioned it here because it is a good thing to do as a starting point. I turned it into a tip.